### PR TITLE
Add support for optional "required by" date for holds.

### PR DESCRIPTION
**Pull request by EreMaijala**
"requiredByDateOptional" can be used when the user needs to be able to choose a last effective date for a hold, but is not required to do so. Additionally "defaultRequiredDate" can now be set to no value or empty string for no default. Support for unspecified "required by" date is included for Demo, KohaRest and SierraRest drivers.

Note that Sierra likely supports fetching `notNeededAfterDate` also before API v6, but I've been unable to verify this, so decided to play safe.

### CODE REVIEW

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @EreMaijala, this looks good to me, but a couple of questions/suggestions before we merge:

1.) Should we add a test for placing an optional hold, so we can differentiate/confirm the behavior in the Mink tests?

2.) Should we create a wiki page to document all the extra hold fields and their meanings? I think we now have enough that it's getting confusing, and it might be helpful to improve documentation. Maybe each ILS driver .ini file can list the supported values and include a link to a wiki page elaborating on the purpose of those values... and then the wiki page can include a note explaining that not every driver supports every option. I'm happy to work on that as a follow-up to this PR, but thought it worth discussing...

File: themes/bootstrap3/templates/record/hold.phtml [hunk 1]
```diff
@@ -31,7 +31,7 @@
     </div>
   <?php endif; ?>
 
-  <?php if (in_array("requiredByDate", $this->extraHoldFields)): ?>
+  <?php if (in_array("requiredByDate", $this->extraHoldFields) || in_array("requiredByDateOptional", $this->extraHoldFields)): ?>
```
Comment (demiankatz) [inline themes/bootstrap3/templates/record/hold.phtml:34]: Do we need to display a different label, or some other visual indicator, to differentiate between optional and non-optional versions of the behavior?

Comment (EreMaijala) [reply themes/bootstrap3/templates/record/hold.phtml:34]: Added!

File: languages/sv.ini [hunk 2]
```diff
@@ -583,6 +583,7 @@ hold_record_already_on_loan = "Du har detta material redan utlånad"
 hold_request_group = "Reservera från"
 hold_requested_group = "Reserverad från"
 hold_required_by = "Sista giltighetsdagen"
+hold_required_by = "Sista giltighetsdagen (valfri)"
```
Comment (demiankatz) [inline languages/sv.ini]: Key is missing _optional here.

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks for adding the label -- just one minor problem!

File: languages/sv.ini (hunk 2 above)
Comment (EreMaijala) [reply languages/sv.ini]: Oops, fixed.

**Review by demiankatz — APPROVED** [review]
Thanks, @EreMaijala, this looks good to me!

**Pull request merged**